library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -1178,17 +1189,22 @@ impl<T, A: Allocator> VecDeque<T, A> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn is_empty(&self) -> bool {
-        self.tail == self.head
+        self.len == 0
     }
 
-    fn range_tail_head<R>(&self, range: R) -> (usize, usize)
+    fn slice_ranges<R>(&self, range: R) -> (Range<usize>, Range<usize>)
```

Comment:
idea: could this return the ranges into `buf` instead?  Then I think the definitions for `as_slices(_mut)` could both call `self.slice_ranges(..)` and two `get_unchecked`, and I think there'd be fewer cases overall, since right now there's two in `as_slices` *and* the three cases in here.

I think if this returns ranges into `buf`, then there are only two possible returns:
- `( self.wrap_index(start)..self.wrap_index(end), 0..0 )`, or
- `( self.wrap_index(start)..self.capacity(), 0..self.wrap_index(end) )`

(If you look at this and it's not actually better, absolutely push back.  I just stared at this method for a while and wasn't quite sure what it was doing for a while.)

Reply:
That sounds like it could work pretty well. It would also make the `Drain` destructor simpler, as right now I have a separate function in there to get the remaining slices of that. I suppose the only sight downside of this would be that it introduces a bit of `unsafe` in the `iter(_mut)()` functions, but it's probably still better than the current implementation.